src/rdf4cpp/rdf/Literal.cpp

```diff
@@ -153,5 +155,225 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
     }
 }
 
+template<typename OpSelect>
+Literal Literal::numeric_binop_impl(OpSelect op_select, Literal const &other, NodeStorage &node_storage) const {
+    if (this->null() || other.null()) {
+        return Literal{};
+    }
+
+    assert(this->handle_.is_literal() && other.handle_.is_literal());
+
+    auto const after_conversion = [&]() -> std::optional<std::tuple<std::string_view, std::any, std::any>> {
+        auto const this_lit_type = this->handle_.literal_backend().datatype_id;
+        auto const other_lit_type = other.handle_.literal_backend().datatype_id;
+
+        if (this_lit_type == other_lit_type) {
+            return std::make_tuple(this->datatype().identifier(), this->value(), other.value());
+        } else {
+            auto const lhs_datatype = this->datatype().identifier();
+            auto const rhs_datatype = other.datatype().identifier();
+
+            auto const equalizer = datatypes::registry::DatatypeRegistry::get_common_type_conversion(
+                    lhs_datatype,
+                    rhs_datatype);
+
+            if (!equalizer.has_value()) {
+                return std::nullopt;
+            }
+
+            return std::make_tuple(equalizer->target_type_iri,
+                                   equalizer->convert_lhs(this->value()),
+                                   equalizer->convert_rhs(other.value()));
+        }
+    }();
+
+    if (!after_conversion.has_value()) {
+        return Literal{}; // datatype mismatch and not in same promotion hierarchy
+    }
+
+    auto const &[res_identifier, lhs_val, rhs_val] = *after_conversion;
+
+    auto const result_entry = datatypes::registry::DatatypeRegistry::get_entry(res_identifier).value().get();
+
+    if (!result_entry.numeric_ops.has_value()) {
```

Comment:
move up because we do not expect type hierarchies where we have non-numeric types which have might be cast to numeric types.